### PR TITLE
Lesson 07 Complete

### PR DESCRIPTION
**Pull request by stuartkershaw**

### CODE REVIEW

File: students/stuartkershaw/lesson07/test_html_render.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+# Step 1
+
+
+def test_type_Element():
+    test_element = Element(content=None)
+    assert isinstance(test_element, Element) is True
+
+
+def test_class_attributes_Element():
+    test_element = Element(content=None)
```
Comment (natasha-aleksandrova) [inline students/stuartkershaw/lesson07/test_html_render.py]: `content` default is None, no need to explicitly create `Element` with that default

File: students/stuartkershaw/lesson07/test_html_render.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+from html_render import H
+from html_render import Meta
+
+
+# Step 1
+
+
+def test_type_Element():
+    test_element = Element(content=None)
+    assert isinstance(test_element, Element) is True
```
Comment (natasha-aleksandrova) [inline students/stuartkershaw/lesson07/test_html_render.py]: this isn't something something that should be included in the test suite.

File: students/stuartkershaw/lesson07/test_html_render.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+
+
+def test_class_attributes_Element():
+    test_element = Element(content=None)
+    assert test_element.tag_name == ''
+    assert test_element.indentation == 4
+
+
+def test_instance_attributes_Element():
```
Comment (natasha-aleksandrova) [inline students/stuartkershaw/lesson07/test_html_render.py]: function name should be all lower case

File: students/stuartkershaw/lesson07/html_render.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        return " " * self.tree_level * self.indentation
+
+    def apply_attributes(self, attrs):
+        self.attrs = attrs
+
+        for key, value in self.attrs.items():
+            self.file_out.write(f" {key}=\"{value}\"")
+
+    def render(self, file_out, cur_ind=""):
+        self.file_out = file_out
```
Comment (natasha-aleksandrova) [inline students/stuartkershaw/lesson07/html_render.py]: there is usually not a good reason to create an instance attribute outside of `__init__` which is why linters will usually complain about that

File: students/stuartkershaw/lesson07/html_render.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+class Html(Element):
+    tag_name = 'html'
+
+    def render(self, file_out, cur_ind=""):
+        self.file_out = file_out
+        self.cur_ind = cur_ind
+
+        self.file_out.write('<!DOCTYPE html>\n')
+
+        Element.render(self, self.file_out, self.cur_ind)
```
Comment (natasha-aleksandrova) [inline students/stuartkershaw/lesson07/html_render.py]: I believe this is old style. You should use `super().render()`

also, `self.file_out` and `self.cur_ind` shouldn't be defined outside of `__init__`

File: students/stuartkershaw/lesson07/html_render.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+class A(Element):
+    tag_name = 'a'
+
+    def __init__(self, link, content):
+        self.link = link
+        self.content = content
+        self.attributes = {'href': self.link}
+
+        Element.__init__(self, content=self.content, **self.attributes)
```
Comment (natasha-aleksandrova) [inline students/stuartkershaw/lesson07/html_render.py]: `super().__init__` call


you don't need any of the instance var definitions there:
```
def __init__(self, link, content):
    super().__init__(content=content, **{'href': link})
````

File: students/stuartkershaw/lesson07/html_render.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+class H(OneLineTag):
+    tag_name = 'h'
+
+    def __init__(self, level, content):
+        self.level = level
+        self.content = content
+        self.attributes = {'level': self.level}
+
+        OneLineTag.__init__(self, content=self.content, **self.attributes)
```
Comment (natasha-aleksandrova) [inline students/stuartkershaw/lesson07/html_render.py]: same comment as above

File: students/stuartkershaw/lesson07/html_render.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+
+class Meta(SelfClosingTag):
+    tag_name = 'meta'
+
+    def __init__(self, **kwargs):
+        self.attributes = kwargs
+
+        SelfClosingTag.__init__(self, **self.attributes)
```
Comment (natasha-aleksandrova) [inline students/stuartkershaw/lesson07/html_render.py]: you don't need the `__init__` definition here

**Comment by natasha-aleksandrova** [comment]
@stuartkershaw make sure to submit on the course website for the grade

**Comment by stuartkershaw** [comment]
@natasha-aleksandrova Thanks for the feedback - I've just submitted on the course website for the grade.

**Comment by stuartkershaw** [comment]
@natasha-aleksandrova @robalford Thanks for the feedback - I've just submitted on the course website for the grade.

File: students/stuartkershaw/lesson07/html_render.py (hunk 4 above)
Comment (stuartkershaw) [reply students/stuartkershaw/lesson07/html_render.py]: Hi @natasha-aleksandrova - I'm underway with patching this up, but I'm having some difficulty imagining how to change this. With the way that `run_html_render.py` passes `file_out` to the render method, I'm not understanding how to shift that to `__init__` without changing `run_html_render.py`... am I just overlooking something there? Thanks -

**Pull request merged**